pkg/client/query/cache/options.go

```diff
@@ -79,3 +79,53 @@ func WithSessionCountCacheClearFn(numSessionsToClearCache uint) func(context.Con
 		return nil
 	}
 }
+
+// WithClaimSettlementCacheClearFn is used to configure cache clearing at claim settlement.
+//
+// This timing is critical for RelayMiningDifficulty caches to ensure suppliers aren't penalized
+// for using outdated difficulty values when submitting proofs that were generated at session start.
+//
+// Cache clearing occurs at the height where claims are settled to:
+//   - Maintain stable difficulty values throughout the proof submission window
+//   - Prevent suppliers from using stale cached difficulty when submitting proofs
+//   - Allow fresh difficulty calculations for the next session cycle
+func WithClaimSettlementCacheClearFn() func(context.Context, depinject.Config, Cache) error {
+	return func(ctx context.Context, deps depinject.Config, cache Cache) error {
+		var blockClient client.BlockClient
+		var sharedParamsCache client.ParamsCache[sharedtypes.Params]
+		var logger polylog.Logger
+
+		// Inject dependencies
+		if err := depinject.Inject(deps, &blockClient, &sharedParamsCache, &logger); err != nil {
+			return err
+		}
+
+		// Open a channel to observe committed blocks and clear the cache at the right time
+		channel.ForEach(
+			ctx,
+			blockClient.CommittedBlocksSequence(ctx),
+			func(ctx context.Context, block client.Block) {
+				// If
+				sharedParams, found := sharedParamsCache.Get()
+				if !found {
+					logger.Debug().Msg("ℹ️ Shared params not found in cache, skipping cache clearing")
+					return
+				}
+
+				// Calculate the height at which claims for the current session will be settled
+				currentHeight := block.Height()
+				currentSessionStartHeight := sharedtypes.GetSessionStartHeight(&sharedParams, currentHeight)
+				sessionEndToProofWindowCloseNumBlocks := sharedtypes.GetSessionEndToProofWindowCloseBlocks(&sharedParams)
+				claimSettlementHeight := currentSessionStartHeight + sessionEndToProofWindowCloseNumBlocks
+
+				// Clear cache when claims are settled to allow fresh difficulty values for the next cycle
+				if currentHeight == claimSettlementHeight {
+					logger.Debug().Msgf("🧹 Clearing cache at claim settlement height: %d", claimSettlementHeight)
+					cache.Clear()
```

Comment:
My only concern with this PR / implementation is that:

1. On line `109` we check if `sharedParams` is available from the `sharedParamsCache`.
2. If it's not -> we do nothing.
3. However, if it is, we clear the whole cache.

From the name of this function, I'd assume the whole cache is cleard each time.

Same issue exists in `WithSessionCountCacheClearFn` today.

Reply:
@Olshansk

TL;DR; Cache clearing is very prone to cyclic dependencies, and the best way to break it is to get `sharedParams` from its cache.

For more context:
1. Session and settlement cache clearing strategies need `sharedParams`
2. Cache clearing strategies are dependency leafs and cannot depend on queriers (e.g. `sharedQuerier`)
3. The only way to get `sharedParams` is from the cache
4. If `sharedParams` are not cached, wait for the next block (until `sharedParams`) are cached before trigger cache clearing checks (it's always available except initially for the `sharedParams` cache clearing)